config/src/waypoint.rs

```diff
@@ -0,0 +1,164 @@
+// Copyright (c) The Libra Core Contributors
+// SPDX-License-Identifier: Apache-2.0
+
+use anyhow::{ensure, format_err, Error, Result};
+use libra_crypto::hash::{CryptoHash, CryptoHasher, HashValue};
+use libra_crypto_derive::CryptoHasher;
+use libra_types::ledger_info::LedgerInfo;
+use libra_types::transaction::Version;
+use libra_types::validator_set::ValidatorSet;
+use serde::{Deserialize, Serialize};
+use std::fmt::{Display, Formatter};
+use std::str::FromStr;
+
+// The delimiter between the version and the hash.
+const WAYPOINT_DELIMITER: char = ':';
+
+/// Waypoint keeps information about the LedgerInfo on a given reconfiguration, which provides an
+/// off-chain mechanism to verify the sync process right after the restart.
+/// At high level, a trusted waypoint verifies the LedgerInfo for a certain epoch change.
+/// For more information, please refer to the Waypoints documentation.
+#[derive(Clone, Debug, Default, Deserialize, PartialEq, Serialize)]
+pub struct Waypoint {
+    /// The version of the reconfiguration transaction that is being approved by this waypoint.
+    version: Version,
+    /// The hash of the chosen fields of LedgerInfo (including the next validator set).
+    value: HashValue,
```

Comment:
Any reason why this is not the entire `LedgerInfo`? Why "chosen fields"? (It's not even clear what these fields are by looking at this.)

Reply:
The comments don't present the full picture. There is a long discussion behind the waypoint format and protocols in the GH issue referenced by this PR.
The reason for not all the fields being hashed by the waypoint is that not all of them are deterministic (e.g., it could be that some consensus specific fields, like `round` are going to be different for different participants).